src/cordova/test.action.mjs

```diff
@@ -0,0 +1,76 @@
+// Copyright 2022 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import url from 'url';
+import os from 'os';
+import minimist from 'minimist';
+import path from 'path';
+import fs from 'fs/promises';
+import rmfr from 'rmfr';
+
+import {getRootDir} from '../build/get_root_dir.mjs';
+import {getXcodebuildDestination} from '../build/get_xcodebuild_destination.mjs';
+import * as runXcodebuild from '../build/run_xcodebuild.mjs';
+
+const APPLE_ROOT = path.join(getRootDir(), 'src', 'cordova', 'apple');
+const APPLE_LIBRARY_NAME = 'OutlineAppleLib';
+
+/**
+ * @description Tests the parameterized cordova binary (ios, macos).
+ *
+ * @param {string[]} parameters
+ */
+export async function main(...parameters) {
+  const {
+    _: [outlinePlatform],
+  } = minimist(parameters);
+
+  if (outlinePlatform !== 'macos' && outlinePlatform !== 'ios') {
+    throw new Error('Testing is only currently supported for Apple platforms.');
+  }
+
+  if (os.platform() !== 'darwin') {
+    throw new Error('Building an Apple binary requires xcodebuild and can only be done on MacOS');
+  }
+
+  const xcodebuildFlags = {
+    scheme: APPLE_LIBRARY_NAME,
+    destination: await getXcodebuildDestination(outlinePlatform),
+    workspace: path.join(APPLE_ROOT, APPLE_LIBRARY_NAME),
+  };
+
+  await runXcodebuild.clean(xcodebuildFlags);
+
+  const xcodebuildTestFlags = {
+    ...xcodebuildFlags,
+    enableCodeCoverage: 'YES',
+    derivedDataPath: path.join(APPLE_ROOT, 'coverage'),
```

Comment:
Use process.env.BUILD_DIR to specify the subdir here

Reply:
The fix here is trivial: just use the BUILD_DIR to define the APPLE_ROOT. But if you really want to output build artifacts to the source tree, please update the clean action to remove the build outputs. You also need to update the .gitignore file.

Yes, I think it's worth having that discussion. Having a clean per workspace is reasonable, but we don't have that. But wouldn't you still have the same issue of having to define a build folder for that workspace? 

I think I don't get your argument. We have a dedicated build output directory in the outline-server and it works fine. We don't have to use Bazel, though we can borrow concepts. I'm not sure what tooling you are referring to. For cordova we can create the project in the build folder as well, we just didn't get to do it. (That would allow us to move away from npm, while keeping npm for Cordova). We also need to consider the non-JS tooling. In the case of this PR it's not even JS, it's Apple/XCode.

Reply:
I like the idea of separating some types of outputs, including test logs. But instead of creating a different top-level dir for each output type, I suggest we add the folders to a top level OUTPUT_DIR instead, so we don't need to know what all the different types are to know what was generated.

I checked what Bazel does: https://bazel.build/remote/output-directories

They have a bazel-out directory, which they call "output directory" or "output base". That's actually the intention of our BUILD_DIR. It was to capture everything that is output, included generated code, and it could include test results too. But perhaps it should be called OUTPUT_DIR to be clearer.

Then we can add a testlogs/ directory under OUTPUT_DIR.

I like to borrow concepts from Bazel because it's been battle tested for over a decade in a monorepo context with multiple languages and platforms. They thought carefully about all those decisions. We can borrow from other systems too.